### PR TITLE
Ml/custom nodes

### PR DESCRIPTION
**Pull request by lubejmatej**
Closes [ch4922, ch3749]

### CODE REVIEW

**Comment by FrederikBolding** [comment]
@lubejmatej Please rebase this before we can review it 😄

File: common/v2/components/MobileNavBar.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  > div {
+    &.tab {
+      position: relative;
+      flex-basis: 0;
+      flex-grow: 1;
+      width: 100%;
+      max-width: 100%;
+
+      padding: 23px 0;
```
Comment (GuillaumeRx) [inline common/v2/components/MobileNavBar.tsx:19]: Could you use `SPACING` styling constant ?

File: common/v2/components/NetworkNodeDropdown.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import NetworkNodeDropdown from './NetworkNodeDropdown';
+import React from 'react';
+import { DEFAULT_NETWORK } from '../config';
+import { storiesOf } from '@storybook/react';
+import AppProviders from '../../AppProviders';
```
Comment (GuillaumeRx) [inline common/v2/components/NetworkNodeDropdown.stories.tsx:5]: Please reorder imports

File: common/v2/components/NetworkNodeDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+import addIcon from 'assets/images/icn-add.svg';
+import editIcon from 'assets/images/icn-edit.svg';
+
+const SContainer = styled.div`
+  display: flex;
+  flex-direction: row;
+  padding: 12px;
+
+  &:hover {
+    background-color: var(--color-gray-lighter);
```
Comment (GuillaumeRx) [inline common/v2/components/NetworkNodeDropdown.tsx:22]: please use TS styling constants with styled components

File: common/v2/components/NetworkNodeDropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+          <Typography value={option.label} />
+          {isFunction(option.onEdit) && isEditEnabled && isCustom === true && (
+            <EditIcon
+              onClick={e => {
+                e.preventDefault();
+
+                if (option.onEdit) {
+                  option.onEdit(option.value);
+                }
+              }}
```
Comment (GuillaumeRx) [inline common/v2/components/NetworkNodeDropdown.tsx:85]: could you extract this function for better readability

File: common/v2/features/Layout/Layout.tsx [hunk 5]
```diff
@@ -42,7 +43,7 @@ const STop = styled.div`
 `;
 
 const SContainer = styled('div')`
-  padding: ${SPACING.BASE};
+  ${({ noPadding }) => (noPadding ? `padding: ${SPACING.BASE} 0;` : `padding: ${SPACING.BASE};`)};
```
Comment (GuillaumeRx) [inline common/v2/features/Layout/Layout.tsx:46]: Maybe it's easier to read with `padding: ${ ({noPadding}) => noPadding ? `${SPACING.BASE} 0` :  ${SPACING.BASE}};` ?

File: common/v2/features/Settings/Settings.tsx [hunk 6]
```diff
@@ -4,13 +4,24 @@ import styled from 'styled-components';
...
+import { NetworkId } from 'v2/types';
+import { CustomNodeConfig } from 'v2/types/node';
+import { DEFAULT_NETWORK, IS_ACTIVE_FEATURE } from 'v2/config';
 
 import settingsIcon from 'common/assets/images/icn-settings.svg';
-import { IS_ACTIVE_FEATURE } from 'v2/config';
+import NetworkNodes from './components/NetworkNodes';
+import AddOrEditNetworkNode from './components/AddOrEditNetworkNode';
+import { AddressBookPanel, AddToAddressBook, GeneralSettings, DangerZone } from './components';
+import MobileNavBar from 'v2/components/MobileNavBar';
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/Settings.tsx:24]: Please add `MobileNavBar` ton `index.ts` in components folder and import it in the import `v2/components`

File: common/v2/components/NetworkNodeDropdown.tsx (hunk 3 above)
Comment (amacar) [reply common/v2/components/NetworkNodeDropdown.tsx:22]: This hover thing can actually be deleted, hover state is now handled in the `Dropdown` component.

File: common/v2/components/MobileNavBar.tsx (hunk 1 above)
Comment (lubejmatej) [reply common/v2/components/MobileNavBar.tsx:19]: No problem, I was just using values from designs.

File: common/v2/features/AddAccount/components/NetworkSelectPanel.tsx [hunk 7]
```diff
@@ -26,23 +29,41 @@ const SButton = styled(Button)`
   bottom: 2em;
 `;
 
+const SLabel = styled.label`
+  margin-top: 15px;
+  margin-bottom: 10px;
```
Comment (GuillaumeRx) [inline common/v2/features/AddAccount/components/NetworkSelectPanel.tsx:34]: Please use `SPACING` constants

File: common/v2/features/AddAccount/components/NetworkSelectPanel.tsx [hunk 8]
```diff
@@ -1,12 +1,15 @@
...
+import { FormData, NetworkId } from 'v2/types';
 import { NetworkSelectDropdown } from 'v2/components';
-import { NetworkContext } from 'v2/services/Store';
+import { NetworkContext, NetworkUtils } from 'v2/services/Store';
+import NetworkNodeDropdown from 'v2/components/NetworkNodeDropdown';
+import { ProviderHandler } from 'v2/services/EthService/network';
+import { ToastContext } from '../../Toasts';
 
 const NetworkForm = styled.div`
   margin-top: 22px;
```
Comment (GuillaumeRx) [inline common/v2/features/AddAccount/components/NetworkSelectPanel.tsx:15]: Can you please update this with `SPACING` constants

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+import { CustomNodeConfig, Network, NetworkId, NodeOptions, NodeType } from 'v2/types';
+import { translateRaw } from 'v2/translations';
+import { DEFAULT_NETWORK, GITHUB_RELEASE_NOTES_URL, LET_ENCRYPT_URL } from 'v2/config';
+import { NetworkUtils } from 'v2/services/Store/Network';
+import { ProviderHandler } from 'v2/services/EthService/network';
+
+import backArrowIcon from 'common/assets/images/icn-back-arrow.svg';
+
+const AddToNetworkNodePanel = styled(DashboardPanel)`
+  padding: 0 30px 15px;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:18]: Please use `SPACING` constants

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+`;
+
+const BackButton = styled(Button)`
+  margin-right: 16px;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:22]: Also some `SPACING` constants

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+`;
+
+const Row = styled.div<{ hidden?: boolean }>`
+  flex-direction: column;
+  ${({ hidden }) => `display: ${hidden ? 'none' : 'flex'};`};
+
+  @media (min-width: ${BREAK_POINTS.SCREEN_XS}) {
+    flex-direction: row;
+    align-items: flex-start;
+    padding-left: 40px;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:32]: `SPACING` constants :)

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+  }
+`;
+
+const SubtitleRow = styled(Row)`
+  margin-top: -25px;
+  padding-bottom: 20px;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:38]: `SPACING` constants :)

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+`;
+
+const Column = styled.div<{ alignSelf?: string }>`
+  flex: 1;
+
+  @media (min-width: ${BREAK_POINTS.SCREEN_SM}) {
+    ${({ alignSelf }) => `align-self: ${alignSelf || 'auto'};`};
+
+    &:not(:first-child) {
+      padding-left: 25px;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:48]: `SPACING` constants :)

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+    }
+  }
+
+  label {
+    padding-left: 3px;
+  }
+`;
+
+const AddressFieldset = styled.fieldset`
+  margin-bottom: 15px;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:62]: `SPACING` constants :)

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+
+  label {
+    display: block;
+    margin-bottom: 9px;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:66]: `SPACING` constants :)

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+  textarea {
+    display: block;
+    width: 100%;
+  }
+`;
+
+const NetworkNodeFieldsButtons = styled.div`
+  display: flex;
+  justify-content: space-between;
+  margin-top: 20px;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:79]: `SPACING` constants :)

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 17]
```diff
@@ -0,0 +1,380 @@
...
+    text-transform: capitalize;
+  }
+
+  @media (min-width: ${BREAK_POINTS.SCREEN_SM}) {
+    margin-top: 0;
+  }
+`;
+
+const SNetworkSelectDropdown = styled(NetworkSelectDropdown)`
+  margin-bottom: 15px;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:90]: `SPACING` constants :)

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 18]
```diff
@@ -0,0 +1,380 @@
...
+`;
+
+const SCheckbox = styled(Checkbox)`
+  margin-bottom: 0;
+  ${({ checked }) => checked && `margin-bottom: 20px;`};
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:95]: `SPACING` constants :)

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 19]
```diff
@@ -0,0 +1,380 @@
...
+  ${({ checked }) =>
+    checked && `@media (min-width: ${BREAK_POINTS.SCREEN_SM}) {margin-bottom: 0;}`};
+`;
+
+const SError = styled.div`
+  margin-top: 10px;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:101]: `SPACING` constants :)

**Comment by FrederikBolding** [comment]
@lubejmatej Is it intentional that selecting `AUTO` doesn't actually select `AUTO` in the dropdown?

File: common/AppRoutes.tsx [hunk 20]
```diff
@@ -29,6 +29,14 @@ const layoutConfig = (path: string) => {
         centered: true,
         paddingV: SPACING.MD
       };
+    case ROUTE_PATHS.SETTINGS.path:
+      return {
+        centered: true,
+        fluid: true,
+        fullW: true,
+        bgColor: '#fff',
```
Comment (GuillaumeRx) [inline common/AppRoutes.tsx:37]: Use `COLORS` constant please

File: common/v2/components/MobileNavBar.tsx [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+
+      &.active {
+        &:after {
+          content: ' ';
+          position: absolute;
+          top: 0;
+          bottom: 0;
+          left: 0;
+          right: 0;
+          background-color: white;
```
Comment (GuillaumeRx) [inline common/v2/components/MobileNavBar.tsx:31]: Use `COLORS` constant please

File: common/v2/features/Settings/components/AddOrEditNetworkNode.tsx [hunk 22]
```diff
@@ -0,0 +1,380 @@
...
+  }
+`;
+
+const AddressFieldset = styled.fieldset`
+  margin-bottom: ${SPACING.SM};
+
+  label {
+    display: block;
+    margin-bottom: ${SPACING.SM};
+    color: #163150;
```
Comment (GuillaumeRx) [inline common/v2/features/Settings/components/AddOrEditNetworkNode.tsx:67]: Use `COLORS` constant please

File: common/v2/theme/constants.ts [hunk 23]
```diff
@@ -85,6 +85,7 @@ export const LINE_HEIGHT = {
 };
 
 export const SPACING = {
+  NONE: '0',
```
Comment (GuillaumeRx) [inline common/v2/theme/constants.ts:88]: I'm not sure this has an utility since 0 will always be 0 :)

Comment (lubejmatej) [reply common/v2/theme/constants.ts:88]: Yeah, good point

File: common/v2/features/Settings/components/AddOrEditNetworkNode.stories.tsx [hunk 24]
```diff
@@ -0,0 +1,84 @@
+import React from 'react';
+import { ThemeProvider } from 'styled-components';
+import { storiesOf } from '@storybook/react';
+
+import { DEFAULT_NETWORK } from 'v2/config';
+import { NETWORKS_CONFIG, NODES_CONFIG } from 'v2/database/data';
+import { CustomNodeConfig, Network, NetworkId, NodeType } from 'v2/types';
+import { GAU_THEME } from 'v2/theme';
+import AppProviders from 'v2/../AppProviders';
```
[the remark this answers is not in the document]
Comment (lubejmatej) [reply common/v2/features/Settings/components/AddOrEditNetworkNode.stories.tsx:9]: Must have been an auto import from IDE

File: common/v2/types/node.ts [hunk 25]
```diff
@@ -16,40 +14,29 @@ export interface NodeConfig {
...
-export interface NodeOptions {
+interface NodeBase {
+  isCustom?: boolean;
   name: string;
   type: NodeType;
   service: string;
   url: string;
-  isCustom?: boolean;
-  isAuto?: boolean;
-  network?: string;
```
[the remark this answers is not in the document]
Comment (lubejmatej) [reply common/v2/types/node.ts:26]: Currently custom node is stored inside network. So we know which network it belongs to. And after refactoring the localstorage, we can just keep it in object with networkId key name. So we dont duplicate the unnecessary data.

File: common/v2/types/network.ts [hunk 26]
```diff
@@ -38,6 +38,8 @@ export interface Network {
   gasPriceSettings: GasPrice;
   shouldEstimateGasPrice?: boolean;
   nodes: NodeOptions[];
+  selectedNode?: string;
+  autoNode?: string;
```
[the remark this answers is not in the document]
Comment (lubejmatej) [reply common/v2/types/network.ts:42]: Yes

File: common/v2/components/MobileNavBar.tsx [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+      width: 100%;
+    }
+  }
+
+  @media (min-width: ${BREAK_POINTS.SCREEN_XS}) {
+    margin-top: ${IS_MOBILE && '-29px'};
+  }
+`;
+
+const MobileNavBar: FC = ({ children }) => {
```
[the remark this answers is not in the document]
Comment (lubejmatej) [reply common/v2/components/MobileNavBar.tsx:56]: Currently there are no props to pass through.

**Comment by lubejmatej** [comment]
> @lubejmatej Is it intentional that selecting `AUTO` doesn't actually select `AUTO` in the dropdown?

It should select Auto node, but localstorage networks should be newly initialized. Try opening in incognito? After refactoring localstorage keys, that is currently in progress, it shoudn't be a problem anymore.

**Comment by lubejmatej** [comment]
> Designs feedback :
> 
> * Where is the possibility to delete a node ?
> * The dropdown is too large, check designs
> * Label of table on top the node row should read "node"
> * Mobile implementation doesn’t respect designs

1. Delete node is inside edit form;
2. I lowered the spacings;
3. This one we actually agreed to be 'Default node', because it uses other nodes, in case the selected one fails. So it made more sense;
4. Should more closely resemble the mobile designs now;

**Pull request merged**